feat: Kotlin 2.4.20-Beta1 release (#5628)
JetBrains/kotlin-web-site: docs/topics/whatsnew/whatsnew-eap.md

The Kotlin %kotlinEapVersion% release is out! Here are some details of this EAP release:

> * **Standard library:** [`StackTraceRecoverable` interface for integration with `kotlinx.coroutines`](#standard-library-new-stacktracerecoverable-interface)

Reviewer: ```suggestion
* **Standard library:** [Add the `StackTraceRecoverable` interface to support coroutine stack trace recovery](#standard-library-new-stacktracerecoverable-interface)
```

Reviewer: or if we want it more neutral, perhaps we could copy the title from below: Introduce the `StackTraceRecoverable` interface in the standard library

> * **Kotlin/Wasm:** [Improved top-level `require()` calls and companion object initialization order](#kotlin-wasm)

Reviewer: ```suggestion
* **Kotlin/Wasm:** [Changes to top-level `require()` calls and improved companion object initialization order](#kotlin-wasm)
```

To update to the new Kotlin version, make sure your IDE is updated to the latest version and [change the Kotlin version](releases.md#update-to-a-new-kotlin-version)
to %kotlinEapVersion% in your build scripts.

> ## Standard library: new `StackTraceRecoverable` interface

Reviewer: ```suggestion
## Standard library: Introduce `StackTraceRecoverable` interface in the standard library 
```

Reviewer: we could probably omit in the standard library, but new makes it feel like it's a new thing, but it's just introduced there as well

// to pass the cause to the IllegalStateException constructor
private constructor(
    val line: Int,
    private val detail: String,
    cause: Throwable?,
) : IllegalStateException("When editing line $line: $detail", cause),
    // Implements StackTraceRecoverable for stack trace recovery
    StackTraceRecoverable<FileEditException> {

    constructor(line: Int, detail: String) : this(line, detail, null)

    // Copies the line number and message details
    override fun copyForStackTraceRecovery(): FileEditException =
        FileEditException(line, detail, this)
}

@OptIn(ExperimentalStdlibCoroutineSupportApi::class) 
fun main() {
    val original = FileEditException(15, "Unexpected token")
    val copy = original.copyForStackTraceRecovery()

    println(copy.message)
    // When editing line 15: Unexpected token

    println(copy.cause == original)
    // true
}
```

For more information, see the feature's [KEEP](https://github.com/Kotlin/KEEP/blob/main/proposals/stdlib/KEEP-0461-stacktrace-recoverable.md).
We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-86595).

## Kotlin/Native: Incremental compilation enabled by default

> Starting with %kotlinEapVersion`, incremental compilation of `klib` artifacts is enabled by default.

Reviewer: ```suggestion
Starting with %kotlinEapVersion%, incremental compilation of `klib` artifacts is enabled by default.
```

> With incremental compilation, if only a part of the `klib` artifact produced by the project module changes, just a part

Reviewer: ```suggestion
With incremental compilation, if only a part of the `klib` artifact produced by the project module changes, only the affected part
```

Reviewer: (suggestion) would this work? 🤔

of the `klib` is further recompiled into a binary.

This optimization was first introduced in [Kotlin 1.9.20](whatsnew1920.md#incremental-compilation-of-klib-artifacts)

> and proved to drastically reduce compilation time for debug builds.

Reviewer: ```suggestion
and has proven to drastically reduce compilation time for debug builds.
```

Reviewer: (suggestion) - the original works as well

```

If your project relies on dependencies that require a top-level `require()` function, add it as a property of
`globalThis` as a workaround:

```kotlin
@JsFun(
    """
    ((module) => {
        globalThis.require = module.default.createRequire(import.meta.url)
        return () => {}
    })(await import("node:module"))
"""
)
external fun defineRequire()
```

If you run into any issues, share your feedback in our [issue tracker](https://youtrack.jetbrains.com/projects/KT/issues/KT-86192).

### Improved companion object initialization order

Kotlin/Wasm now initializes superclass companion objects before subclass companion objects, matching the JVM behavior.
Previously, the initialization could be reversed, leading to inconsistent behavior across platforms.

The update improves cross-platform consistency and reduces platform-specific differences in class initialization behavior.
It also enables correct handling of companion object initialization in deeper inheritance hierarchies, including cases
where intermediate classes don’t declare companion objects.

## Kotlin/JS: new DSL for browser-testing
<primary-label ref="experimental-opt-in"/>

> Kotlin %kotlinEapVersion` introduces a new experimental DSL for running Kotlin/JS tests in a browser environment.

Reviewer: ```suggestion
Kotlin %kotlinEapVersion% introduces a new experimental DSL for running Kotlin/JS tests in a browser environment.
```

Currently, the Kotlin Gradle plugin uses [Karma](https://github.com/karma-runner/karma) as a browser launcher to run
JavaScript tests across different browsers. The Karma project has been deprecated for 2 years now, which made us explore
alternative ways to support browser testing.

The new DSL is intended to replace Karma as a manager of different tools under the hood and includes:

* [Mocha](https://mochajs.org/) as a test runner.
* [Webpack](https://webpack.js.org/) as a bundler (will be replaced with [Vite](https://vite.dev/)
  in [future releases](https://youtrack.jetbrains.com/issue/KT-48308/)).
* [Playwright](https://playwright.dev/) as a browser driver and a distribution manager that supports Chromium, Firefox,
  and WebKit (Safari) browser engines.

> To try out the new testing DSL, and the opt-in `test{}` block inside `browser{}` for your Kotlin/JS target:

Reviewer: Sanity check: shouldn't this be 
To try out the new testing DSL, add the opt-in `test{}` block

```kotlin
kotlin {
    js {
        browser {
            @OptIn(ExperimentalJsTestDsl::class)
            // Add and configure the new test{} block
            test {
                // Set up options common for all browsers
                browserDefaults {
                    timeout = Duration.ofSeconds(2)
                    headless = true
                }
                // Enable Chromium test runner
                chromium {
                    // Override the common timeout option
                    timeout = Duration.ofSeconds(5)
                    launchArgs.add("--no-sandbox")
                }
                // Enable Firefox test runner
                firefox()
                // Enable WebKit test runner
                webkit { }

>                 // Enable and configure additional WebKit test runner

Reviewer: ```suggestion
                // Enable and configure an additional WebKit test runner
```

                webkit("noheadless") {
                    // Set up custom options
                    headless = false
                }
            }
        }
    }
}
```

The new DSL is in active development. We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-66897).

## Build tools API: Support for Kotlin/JS, Kotlin/Wasm, and Kotlin metadata
<primary-label ref="experimental-general"/>

> In [Kotlin 2.2.20](whatsnew22.md#new-experimental-build-tools-api), the build tools API (BTA) became available for

Reviewer: ```suggestion
In [Kotlin 2.2.0](whatsnew22.md#new-experimental-build-tools-api), the build tools API (BTA) became available for
```

Kotlin/JVM. Kotlin 2.4.20-Beta1 takes the next step toward BTA stabilization by adding support for new targets:
Kotlin/JS, Kotlin/Wasm, and Kotlin metadata.

This makes the Kotlin Gradle plugin interact with the compiler more consistently. In some cases, you can also benefit
from faster, more stable compilation.

The BTA is a universal API that acts as an abstraction layer between build systems and the Kotlin compiler ecosystem.
It helps support Kotlin features and compatibility with the Kotlin compiler in available build tools.

We plan to roll out the BTA support for the new targets in the Kotlin Gradle plugin gradually:

* In Kotlin 2.4.20-Beta1, BTA is enabled in Kotlin/JS, Kotlin/Wasm, and Kotlin metadata by default to gather feedback.
  No additional changes in projects are required.

> * In Kotlin 2.4.20-Beta2−2.4.20, BTA in the new targets will be available as an opt-in. To try it out, add the

Reviewer: hmm 🤔 
perhaps we could phrase this as simply:

Starting with Kotlin 2.4.20-Beta2 - or are we planning on changing it immediately after 2.4.20? 🤔  (I mean immediately in 2.5.0-beta1)

 in that case perhaps something like: From Kotlin 2.4.20-Beta2 through the Kotlin 2.4.20 release,
